src/engine/loopingcontrol.cpp

```diff
@@ -711,151 +820,280 @@ void LoopingControl::clearActiveBeatLoop() {
     }
 }
 
-void LoopingControl::slotBeatLoop(double beats, bool keepStartPoint) {
-    int samples = m_pTrackSamples->get();
-    if (!m_pTrack || samples == 0) {
-        clearActiveBeatLoop();
-        return;
+bool LoopingControl::currentLoopMatchesBeatloopSize() {
+    LoopSamples loopSamples = m_loopSamples.getValue();
+    int currentLoopLength = loopSamples.end - loopSamples.start;
+
+    double beatloopSize = m_pCOBeatLoopSize->get();
+    int beatLoopLength = calculateEndOfBeatloop(loopSamples.start, beatloopSize)
+                         - loopSamples.start;
+
+    // qDebug() << "LoopingControl::currentLoopMatchesBeatloopSize" << currentLoopLength << beatLoopLength;
+
+    // calculateEndOfBeatloop will round to the nearest even sample, so allow
+    // 2 samples of wiggle room.
+    return abs(currentLoopLength - beatLoopLength) <= 2;
+}
+
+int LoopingControl::calculateEndOfBeatloop(int startSample, double beatloopSize) {
+    if (m_pBeats == nullptr) {
+        return kNoTrigger;
     }
 
-    if (!m_pBeats) {
-        clearActiveBeatLoop();
-        return;
+    int fullbeats = static_cast<int>(beatloopSize);
```

Comment:
Why we have a int roundtrip here? 
We are working towards a floating point transport engine, so the start sample and the beatloopSize can have a fractional value.